### PR TITLE
Cache pip and extension compilation

### PR DESCRIPTION
**Pull request by thedrow**
Since we constantly compile code when testing and the tests rarely change,
it'd be wise to setup ccache for them.

I've also went ahead and cached pypi packages for good measure.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -69,12 +69,35 @@ jobs:
...
   steps:
   - task: UsePythonVersion@0
     inputs:
       versionSpec: '$(python.version)'
     displayName: 'Use Python $(python.version)'
 
+  - script: |
+      sudo apt update && sudo apt install -y ccache
+      echo "##vso[task.prependpath]/usr/lib/ccache"
+    displayName: Install ccache and update PATH to use linked versions of gcc, cc, etc
```
Comment (antocuni) [inline azure-pipelines.yml:84]: where is the code which update the `PATH`? I see only a `echo` command here

File: azure-pipelines.yml [hunk 2]
```diff
@@ -69,12 +69,35 @@ jobs:
...
+  - task: Cache@2
+    inputs:
+      key: 'ccache | "$(Agent.OS)" | "$(python.version)"'
+      path: $(CCACHE_DIR)
+    displayName: Cache C compilation with ccache
+
+  - task: Cache@2
+    inputs:
+      key: 'python | "$(Agent.OS)"'
+      restoreKeys: |
```
Comment (antocuni) [inline azure-pipelines.yml:95]: this is more out of curiosity than a review, but why do we need `restoreKeys` for the pip cache and not for the ccache?

File: azure-pipelines.yml (hunk 1 above)
Comment (thedrow) [reply azure-pipelines.yml:84]: The echo command prepends it to the path.
This is copied from azure pipelines' documentation.

File: azure-pipelines.yml (hunk 2 above)
Comment (thedrow) [reply azure-pipelines.yml:95]: I didn't give this much thought.
I just copied it from their documentation.

**Comment by thedrow** [comment]
As you can see the PoC tests' runtime has been cut by 50%.
It also cuts our test suite's runtime by 50%.

You can try this locally by installing ccache and running the test suite twice while using it.

**Comment by hodgestar** [comment]
Continuing in #99 so that I can update it with the recent changes to master.

**Pull request closed without merging**